### DISCUSSION TITLE
Overriding Abstract Classes in Pimcore

### DISCUSSION THREAD

**Discussion by ChakibBrikciSid** [pimcore/pimcore#16979] [Q&A / Support] [answered]
Hello,

I'm currently in the process of extending and customizing Pimcore's Cart management system. As part of this customization, I've developed a new abstract class named `AbstractCart`, which extends `Pimcore\Bundle\EcommerceFrameworkBundle\CartManager\AbstractCart`.

Upon reviewing the documentation, I came across a note stating that "you cannot override abstract model classes directly". Specifically, it mentions:

"But you cannot override Pimcore\Model\Asset (or the other abstract model classes) ifself (as it is the parent class of for example Pimcore\Model\Asset\Image and that would mean to change the class hierarchy)."  [Documentation link](https://pimcore.com/docs/platform/2023.2/Pimcore/Extending_Pimcore/Overriding_Models/)

- I'm seeking clarification on what this means in the context of overriding abstract classes.

- Basically we want to find a way to add new shared methods to `Cart` and `SessionCart` classes. I would also take any other option than overriding `AbstractCart`, the goal is to `avoid duplicated code` in `Cart` and `SessionCart`, as the methods would be the same.

Any insights or best practices would be greatly appreciated.

Thank you!

**Comment by markus-moser** [comment, marked as answer]
@ChakibBrikciSid You cannot change the parent abstract class of the concrete cart implementations as this is directly done with PHP's inheritance feature (therefore no DI possible). If you need to add additional methods to both cart implementations use a trait (and eventually add a additional interface).

**Reply by ChakibBrikciSid** [reply]
Hello @markus-moser, 

It seems like using a trait could be a convenient solution for addressing our issue of duplicate code.
Once I jump back to the story, I'll proceed with implementing the trait.

Thanks for your input, and I hope you have a great day!
